internal/cluster/certificates.go

```diff
@@ -0,0 +1,408 @@
+}
+
+// TODO: consider moving a generating function into the Certificate object itself?
+type certGenerator func() (*certs.KeyPair, error)
+
+// Generate will generate any certificates that do not have KeyPair data.
+func (c Certificates) Generate() error {
+	for _, certificate := range c {
+		if certificate.KeyPair == nil {
+			var generator certGenerator
+			switch certificate.Name {
+			case ServiceAccountName:
+				generator = generateServiceAccountKeys
+			default:
+				generator = generateCACert
+			}
+
+			kp, err := generator()
+			if err != nil {
+				return err
+			}
+			certificate.KeyPair = kp
+			certificate.Generated = true
+		}
+	}
+	return nil
+}
+
+// SaveGenerated will save any certificates that have been generated as Kubernetes secrets.
+func (c Certificates) SaveGenerated(ctx context.Context, ctrlclient client.Client, cluster *clusterv1.Cluster, config *bootstrapv1.KubeadmConfig) error {
+	for _, certificate := range c {
+		if !certificate.Generated {
+			continue
+		}
+		secret := certificate.AsSecret(cluster, config)
+		if err := ctrlclient.Create(ctx, secret); err != nil {
+			return errors.WithStack(err)
+		}
+	}
+	return nil
+}
+
+// GetOrCreate is a convenience function that wraps cluster bootstrap certificate behavior.
+func (c Certificates) GetOrCreate(ctx context.Context, ctrlclient client.Client, cluster *clusterv1.Cluster, config *bootstrapv1.KubeadmConfig) error {
+	// Get the certificates that exist
+	if err := c.Lookup(ctx, ctrlclient, cluster); err != nil {
+		return err
+	}
+
+	// Create the certificates that don't exist
+	if err := c.Generate(); err != nil {
+		return err
+	}
+
+	// Save any certificates that have been generated
+	if err := c.SaveGenerated(ctx, ctrlclient, cluster, config); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Certificate represents a single certificate CA.
+type Certificate struct {
+	Generated bool
+	Name      string
+	KeyPair   *certs.KeyPair
+}
+
+// SecretName is the expected name of the Kubernetes secret the certificate is saved in.
+func (c *Certificate) SecretName(clusterName string) string {
+	return fmt.Sprintf("%s-%s", clusterName, c.Name)
+}
+
+// Hashes hashes all the certificates stored in a CA certificate.
+func (c *Certificate) Hashes() ([]string, error) {
+	certificates, err := cert.ParseCertsPEM(c.KeyPair.Cert)
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to parse %s certificate", c.Name)
+	}
+	out := make([]string, 0)
+	for _, c := range certificates {
+		out = append(out, hashCert(c))
+	}
+	return out, nil
+}
+
+// hashCert calculates the sha256 of certificate.
+func hashCert(certificate *x509.Certificate) string {
+	spkiHash := sha256.Sum256(certificate.RawSubjectPublicKeyInfo)
+	return "sha256:" + strings.ToLower(hex.EncodeToString(spkiHash[:]))
+}
+
+// AsSecret converts a single certificate into a Kubernetes secret.
+func (c *Certificate) AsSecret(cluster *clusterv1.Cluster, config *bootstrapv1.KubeadmConfig) *corev1.Secret {
```

Comment:
We certainly could. Would it be ok to do this in a follow up issue so we can discuss behavior expectations?

It might be worth mentioning this package is internal so no one outside of CABPK will be able to use it, but I'm sure it would still be more friendly to return an error or produce an empty secret.

Reply:
Sure things!